pkg/node-servant/static-pod-upgrade/upgrade.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2023 The OpenYurt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/pkg/node-servant/static-pod-upgrade/util"
+)
+
+const (
+	DefaultUpgradeDir                   = "openyurtio-upgrade"
+	DefaultStaticPodRunningCheckTimeout = 2 * time.Minute
+
+	// TODO: use constant value of static-pod controller
+	OTA  = "ota"
+	Auto = "auto"
+)
+
+var (
+	DefaultConfigmapPath = "/data"
+	DefaultManifestPath  = "/etc/kubernetes/manifests"
+)
+
+type Controller struct {
+	// Name of static pod
+	name string
+	// Namespace of static pod
+	namespace string
+	// Manifest file name of static pod
+	manifest string
+	// The latest static pod hash
+	hash string
+	// Only support `OTA` and `Auto`
+	upgradeMode string
+
+	// Manifest path of static pod, default `/etc/kubernetes/manifests/manifestName.yaml`
+	manifestPath string
+	// The backup manifest path, default `/etc/kubernetes/manifests/openyurtio-upgrade/manifestName.bak`
+	bakManifestPath string
+	// Default is `/data/podName`
+	configMapDataPath string
+	// The latest manifest path, default `/etc/kubernetes/manifests/openyurtio-upgrade/manifestName.upgrade`
+	upgradeManifestPath string
+}
+
+func New(name, namespace, manifest, hash, mode string) (*Controller, error) {
+	ctrl := &Controller{
+		name:        name,
+		namespace:   namespace,
+		manifest:    manifest,
+		hash:        hash,
+		upgradeMode: mode,
+	}
+
+	ctrl.manifestPath = filepath.Join(DefaultManifestPath, util.WithYamlSuffix(ctrl.manifest))
+	ctrl.bakManifestPath = filepath.Join(DefaultManifestPath, DefaultUpgradeDir, util.WithBackupSuffix(ctrl.manifest))
+	ctrl.configMapDataPath = filepath.Join(DefaultConfigmapPath, ctrl.manifest)
+	ctrl.upgradeManifestPath = filepath.Join(DefaultManifestPath, DefaultUpgradeDir, util.WithUpgradeSuffix(ctrl.manifest))
```

Comment:
I think we should not pollute `/etc/kubernetes/manifests` directory, so how about place `manifest.bak` and `manifests.upgrade` file under `/tmp/manifests` directory?

Reply:
done